src/VisualStudio/VisualStudioDiagnosticsToolWindow/Loggers/OutputWindowLogger.cs

```diff
@@ -0,0 +1,157 @@
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.CodeAnalysis.Internal.Log
+{
+    /// <summary>
+    /// Implementation of <see cref="ILogger"/> that output to output window
+    /// </summary>
+    internal sealed class OutputWindowLogger : ILogger
+    {
+        private readonly Func<FunctionId, bool> _loggingChecker;
+
+        public OutputWindowLogger()
+            : this((Func<FunctionId, bool>)null)
+        {
+        }
+
+        public OutputWindowLogger(IGlobalOptionService optionService)
+            : this(Logger.GetLoggingChecker(optionService))
+        {
+        }
+
+        public OutputWindowLogger(Func<FunctionId, bool> loggingChecker)
+        {
+            _loggingChecker = loggingChecker;
+        }
+
+        public bool IsEnabled(FunctionId functionId)
+        {
+            return _loggingChecker == null || _loggingChecker(functionId);
+        }
+
+        public void Log(FunctionId functionId, LogMessage logMessage)
+        {
+            OutputPane.WriteLine(string.Format("[{0}] {1} - {2}", Thread.CurrentThread.ManagedThreadId, functionId.ToString(), logMessage.GetMessage()));
+        }
+
+        public void LogBlockStart(FunctionId functionId, LogMessage logMessage, int uniquePairId, CancellationToken cancellationToken)
+        {
+            OutputPane.WriteLine(string.Format("[{0}] Start({1}) : {2} - {3}", Thread.CurrentThread.ManagedThreadId, uniquePairId, functionId.ToString(), logMessage.GetMessage()));
+        }
+
+        public void LogBlockEnd(FunctionId functionId, LogMessage logMessage, int uniquePairId, int delta, CancellationToken cancellationToken)
+        {
+            var functionString = functionId.ToString() + (cancellationToken.IsCancellationRequested ? " Canceled" : string.Empty);
+            OutputPane.WriteLine(string.Format("[{0}] End({1}) : [{2}ms] {3}", Thread.CurrentThread.ManagedThreadId, uniquePairId, delta, functionString));
+        }
+
+        private class OutputPane
+        {
+            private static readonly Guid s_outputPaneGuid = new Guid("BBAFF416-4AF5-41F2-9F93-91F283E43C3B");
+
+            public static readonly OutputPane s_instance = new OutputPane();
+
+            private readonly IServiceProvider _serviceProvider;
+
+            public static void WriteLine(string value)
+            {
+                s_instance.WriteLineInternal(value);
+            }
+
+            public OutputPane()
+            {
+                _serviceProvider = ServiceProvider.GlobalProvider;
+            }
+
+            private IVsOutputWindowPane _doNotAccessDirectlyOutputPane;
+
+            private void WriteLineInternal(string value)
+            {
+                var pane = GetPane();
+                if (pane == null)
+                {
+                    return;
+                }
+
+                pane.OutputStringThreadSafe(value + Environment.NewLine);
```

Comment:
May want to consider batching if there are too many of these events firing continuously.

Reply:
this is off by default and only get enabled when someone explicitly enable for investigation. so I think simpler form should be fine for now.